### DISCUSSION TITLE
How to pass CMake options to dependency

### DISCUSSION THREAD

**Discussion by HispidaBytes** [microsoft/vcpkg#26050] [Q&A]
Hi,

I currently have a project which consists of a shared library that has a dependency on the spdlog library. spdlog is linked statically to my shared library.
In my current configuration (without vcpkg) I'm building the spdlog library with the following CMake options, required for my project:

`-DSPDLOG_PREVENT_CHILD_FD=ON -DCMAKE_POSITION_INDEPENDENT_CODE=ON`

The first option is required by the internal logic of my shared library.
The second one because the spdlog static lib will be used inside a shared library (see this: https://github.com/gabime/spdlog/issues/2226). If I don't add this one, the shared library fails to build.

I'm unable to understand how I can enable those options when using vcpkg to resolve dependencies for my project.
Do I have to create a custom port and use `--overlay-ports` every time ?
Can someone please help me with this ?

Thank you

**Comment by JackBoosY** [comment]
> Do I have to create a custom port and use --overlay-ports every time ?

Yes. Or you can make a PR to add this feature if it's official.

**Reply by dg0yt** [reply]
> Do I have to create a custom port and use --overlay-ports every time ?

You can use custom triplet files with
~~~
if(PORT STREQUAL "spdlog")
    set(VCPKG_CMAKE_CONFIGURE_OPTIONS -DSPDLOG_PREVENT_CHILD_FD=ON -DCMAKE_POSITION_INDEPENDENT_CODE=ON)
endif()
~~~
(vcpkg should already request PIC via toolchain flags.)

**Reply by netcorefan1** [reply]
@dg0yt This saved me from opening a new discussion! I was starting to believe that it was really needed to deal with custom ports just to pass two flags. I tried your suggestion:
```
set(VCPKG_TARGET_ARCHITECTURE x64)
if(${PORT} MATCHES "zlib-ng")
	set(VCPKG_CRT_LINKAGE static)
	set(VCPKG_LIBRARY_LINKAGE static)
	set(VCPKG_CMAKE_CONFIGURE_OPTIONS -DCMAKE_POSITION_INDEPENDENT_CODE=ON -DZLIB_COMPAT=ON)
else()
	set(VCPKG_CRT_LINKAGE dynamic)
	set(VCPKG_LIBRARY_LINKAGE dynamic)
endif()
```
The out.log of my custom triplet correctly shows the arguments (`"-DCMAKE_POSITION_INDEPENDENT_CODE=ON" "-DZLIB_COMPAT=ON"`) and I can see the feature enabled:
```
-- The following features have been enabled:

 * CMAKE_BUILD_TYPE, Build type: Release (selected)
 * WITH_GZFILEOP, Compile with support for gzFile related functions
 * ZLIB_COMPAT, Compile with zlib compatible API
```
However, the resulting output is the same even without VCPKG_CMAKE_CONFIGURE_OPTIONS.
To the left zlib-ng compiled from source with `cmake -DZLIB_COMPAT=ON -DCMAKE_POSITION_INDEPENDENT_CODE=ON .` To the right vcpkg.

![zlib](https://user-images.githubusercontent.com/63934725/231270873-501d9ae5-815a-4d0d-9e06-fe7eac3105d0.jpg)

Should I report this to @saucecontrol ? Or I'm  missing something obvious?

**Reply by Neumann-A** [reply]
>  Or I'm missing something obvious?

What do you think is the problem ?

**Reply by netcorefan1** [reply]
I have no idea, but I see that the last [update](https://github.com/microsoft/vcpkg/tree/master/ports/zlib-ng) to the port is around 8 months ago and since I compiled from a fresh cloned repo, I'm not sure if in the meantime there have been changes that generate a different output.
I could be perfectly wrong, but I suspect that in vcpkg, zlib and zlib-ng are both statically compiled while in the generated VS solution I get from the source, I can see that zlib is configured as dynamic lib and zlib-ng as static lib. Vcpkg generated ninja projects (for which I have no experience at all) instead of a VS solution, therefore I cannot do a comparison.
From my side there are no differences in configurations of both builds: same default parameters and same `-DZLIB_COMPAT/-DCMAKE_POSITION_INDEPENDENT_CODE` flags.

**Edit**: Well, given that with or without flags, I get the same `zlibstatic.lib` with no changes in size, it could be that zlib project is not even compiled as it should.

**Reply by Neumann-A** [reply]
a) vcpkg builds only one type of library type either static or dynamic
b) from the names upstream uses i assume the default build is both
c) in your triplet you specify a static library so you get the static library only
d) the size is different since CRT linkage probably is different. Rule of thumb: don't mix CRT linkage, switch that back to dynamic. 
e) there is no problem with the port however using it as a direct replacement for zlib will be a bit more involved due to the static library suffix. You could simply copy paste the file and rename it to just zlib.

**Reply by netcorefan1** [reply]
Many thanks for your help. Thanks for the rule of thumb! I switched back to dynamic and I will never do such mistake again. However, size didn't changed. I am surprised to see that this time the fix is easy like renaming `zlibstatic.lib` into `zlib.lib`, but I didn't found a way on how to this programmatically from within the process. Could you point me to the right direction? 

I'm building mainly for OpenCv (but I plan to use it everywhere) and I started by following the suggestions given [here](https://github.com/opencv/opencv/issues/22573#issuecomment-1263508538). They said to build statically to avoid runtime issues although I would have preferred a dynamic dll in order to take advantage of delay loading.
I encountered a couple of times runtime issues (mostly [Attempted to read or write protected memory](https://github.com/microsoft/vcpkg/discussions/30559#discussioncomment-5503607)) when building vcpkg dynamic libs. This occurs especially when I try to replace submodules in repos (which are often old) with new vcpkg updated versions. When I build statically such problems does not appears (or at least I have not yet experienced any).
I really need another rule of thumb for this because it is giving me very bad times. I don't like the idea to load the whole static lib in memory even when I don't use any of its functions, but I don't even want unpredictable runtime issues when I call an invalid function.

With `VCPKG_CRT_LINKAGE`  and `VCPKG_LIBRARY_LINKAGE ` set to dynamic I get this:
![zlib2](https://user-images.githubusercontent.com/63934725/231312544-086504f0-5a4c-4336-b74b-8009af9e2872.jpg)

`zlibstatic.lib` is missing and by looking at the sizes, I'm not sure it contains zlib-ng compiled code.

**Reply by HispidaBytes** [reply]
> Yes. Or you can make a PR to add this feature if it's official.

Yes this is an official option of spdlog.
However I think a PR to add this feature to the spdlog port might break this rule:
https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#do-not-use-features-to-implement-alternatives

Am I right?
